chain/export/export.go

```diff
@@ -0,0 +1,267 @@
+		min:   0,
+		incMg: true,
+		incRt: true,
+		incSt: true,
+	}
+	return ce
+}
+
+func (ce *ChainExporter) Export(ctx context.Context, opts ...ExportOption) error {
+	for _, opt := range opts {
+		opt(ce)
+	}
+	log.Infow("starting chain export", "head", ce.head, "messages", ce.incMg, "receipts", ce.incRt, "stateroots", ce.incSt, "final-epoch", ce.min)
+	defer log.Info("chain export complete")
+
+	var bar *pb.ProgressBar
+	if ce.withProg {
+		bar = pb.New64(int64(ce.head.Height()))
+		bar.ShowTimeLeft = true
+		bar.ShowPercent = true
+		bar.Prefix("epochs")
+		bar.Units = pb.U_NO
+		bar.Start()
+		defer bar.Finish()
+	}
+
+	seen := cid.NewSet()
+	todo := NewStack()
+	h := &car.CarHeader{
+		Roots:   ce.head.Cids(),
+		Version: 1,
+	}
+
+	if err := car.WriteHeader(h, ce.out); err != nil {
+		return err
+	}
+	for _, c := range ce.head.Cids() {
+		todo.Push(c, BlockHeader)
+	}
+
+	for todo.Len() > 0 {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+
+		c, kind := todo.Pop()
+		if !seen.Visit(c) {
+			continue
+		}
+		if c.Prefix().Codec != cid.DagCBOR {
+			continue
+		}
+
+		// we haven't visited this cid and its part of a dag, write it to file
+		if err := ce.writeContent(c); err != nil {
+			return err
+		}
+
+		if kind == BlockHeader {
+			blk, err := ce.store.Get(c)
+			if err != nil {
+				return err
+			}
+
+			var b types.BlockHeader
+			if err := b.UnmarshalCBOR(bytes.NewBuffer(blk.RawData())); err != nil {
+				return xerrors.Errorf("unmarshaling block header (cid=%s): %w", blk, err)
+			}
+
+			if ce.withProg {
+				bar.Set64(int64(ce.head.Height() - b.Height))
+			}
+
+			for _, parent := range b.Parents {
+				if b.Height > 0 {
+					todo.Push(parent, BlockHeader)
+				} else {
+					todo.Push(parent, Dag)
+				}
+			}
+
+			// ensure the genesis state is always exported while skipping export for any state out
+			// of range.
+			if b.Height == 0 || b.Height > ce.min {
+				if ce.incMg {
+					if err := ce.pushNewLinks(b.Messages, todo); err != nil {
+						return err
+					}
+				}
+
+				if ce.incRt {
+					if err := ce.pushNewLinks(b.ParentMessageReceipts, todo); err != nil {
+						return err
+					}
+				}
+
+				if ce.incSt {
+					if err := ce.pushNewLinks(b.ParentStateRoot, todo); err != nil {
+						return err
+					}
+				}
+			}
+		} else if kind == Dag {
+			if err := ce.pushNewLinks(c, todo); err != nil {
+				return err
+			}
+		} else {
+			panic("received undefined cid kind")
+		}
+	}
+	return nil
+}
+
+func (ce *ChainExporter) pushNewLinks(c cid.Cid, s *Stack) error {
+	s.Push(c, Dag)
+	data, err := ce.store.Get(c)
+	if err != nil {
+		return err
+	}
+	return cbg.ScanForLinks(bytes.NewReader(data.RawData()), func(visit cid.Cid) {
+		s.Push(visit, Dag)
+	})
+}
+
+func (ce *ChainExporter) writeContent(c cid.Cid) error {
+	blk, err := ce.store.Get(c)
+	if err != nil {
+		return err
+	}
+	return util.LdWrite(ce.out, c.Bytes(), blk.RawData())
+}
+
+type ExportOption func(ce *ChainExporter)
```

Comment:
I think I'd rather see a single ExportOptions type with the various options as fields. That would be consistent with the rest of the codebase and a lot less code. Especially for something that is only used once.